govgen-collect-gentxs.py

```diff
@@ -0,0 +1,135 @@
+    genesis = load_json(genesis_path)
+    accounts = genesis.get('app_state', {}).get('auth', {}).get('accounts', [])
+    balances = genesis.get('app_state', {}).get('bank', {}).get('balances', [])
+    total_supply = int(genesis['app_state']['bank']['supply'][0]['amount'])
+    if genesis['app_state']['bank']['supply'][0]['denom'] != denom:
+        invalid_denom = genesis['app_state']['bank']['supply'][0]['denom']
+        raise Exception(f'invalid denom {invalid_denom} found in bank supply')
+    if len(genesis['app_state']['bank']['supply']) > 1:
+        raise Exception('supply for GovGen is supposed to only have one coin')
+
+    # dictionary to quickly check if an account exists and its balance
+    accounts_dict = {acc['address']: acc for acc in balances}
+    for acc in accounts:
+        if acc['address'] not in accounts_dict:
+            raise Exception(f'account {acc["address"]} exists in genesis but '
+                            'does not have balance')
+
+    for gentx_name in os.listdir(gentxs_folder):
+        if not gentx_name.endswith('.json'):
+            continue
+
+        gentx_path = os.path.join(gentxs_folder, gentx_name)
+        gentx = load_json(gentx_path)
+
+        delegator_address = gentx['body']['messages'][0]['delegator_address']
+
+        # sanity check
+        if not check_gentx_compliance(gentx):
+            # skip validator ad it did not comply with requirements
+            continue
+
+        # Find the index in the balances list
+        balance_index = next((i for i, item in enumerate(balances) if item["address"] == delegator_address), None)
+
+        if balance_index is not None:
+            # account exists, check if it needs topping off
+            account_balance_info = balances[balance_index]
+            account_balance = int(account_balance_info['coins'][0]['amount'])
+
+            if account_balance_info['coins'][0]['denom'] != denom:
+                invalid_denom = account_balance_info['coins'][0]['denom']
+                raise Exception(f'invalid denom {invalid_denom} found for address {delegator_address}')
+            if len(account_balance_info['coins']) > 1:
+                raise Exception(f'genesis balance for address {delegator_address} expects only one coin type.')
+
+            if account_balance < initial_balance:
+                balances[balance_index] = new_account_balance(delegator_address, denom, initial_balance)
+                top_off_amount = initial_balance - account_balance
+                total_supply += top_off_amount
+        else:
+            new_account = new_genesis_account(delegator_address)
+            balance = new_account_balance(delegator_address, denom, initial_balance)
+            accounts.append(new_account)
+            balances.append(balance)
+            total_supply += initial_balance
+            accounts_dict[delegator_address] = new_account
+
+        genesis['app_state']['genutil']['gen_txs'].append(gentx)
```

Comment:
So we don't need to call `govgend collect-gentxs` right ? 

Are you sure there aren't other validations that this command ensures that are not present in your script?

Reply:
right, because the code here does exactly the same. You can also do a `govgend validate-genesis` generated with this and it should say it's valid.

You could also (if you want to be 100% sure) cross check with what the `collect-gentxs` command does but this should be all.

Reply:
I've checked the collect-gentxs code, and there's some other stuff that might be important:

1) append the peer IP address to config.toml `persistent_peers` field (is this file also something we need to share to all validators like the genesis ?) (see https://github.com/cosmos/cosmos-sdk/blob/23359a0b6d91450f8eda2060818d07cdc8632fd1/x/genutil/collect.go#L37-L38)

2) verify that the validator address, once convert to an account address, matches something in the balances (see https://github.com/cosmos/cosmos-sdk/blob/23359a0b6d91450f8eda2060818d07cdc8632fd1/x/genutil/collect.go#L157-L164)

Reply:
thanks! I'll work on it

Reply:
I have addressed both things. The script now checks for addresses consistency and collects peer info and prints them at the end of execution, so we can use it.